### PR TITLE
Feat(bigquery): add support for FOR .. IN statement

### LINKED ISSUE #2506: BigQuery dialect parse error on "FOR... IN(...) DO"
version 19.0.1
steps to reproduce:
```python
sqlglot.parse(
  "FOR record IN (SELECT word FROM shakespeare) DO SELECT record.word; END FOR;",
  sqlglot.dialects.BigQuery
)
```

<details>
<summary>traceback</summary>

>>> sqlglot.parse("FOR record IN (SELECT word FROM shakespeare) DO SELECT record.word; END FOR;", sqlglot.dialects.BigQuery)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/__init__.py", line 86, in parse
    return dialect.parse(sql, **opts)
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/dialects/dialect.py", line 311, in parse
    return self.parser(**opts).parse(self.tokenize(sql), sql)
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 987, in parse
    return self._parse(
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 1053, in _parse
    expressions.append(parse_method(self))
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 1242, in _parse_statement
    expression = self._parse_expression()
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 3232, in _parse_expression
    return self._parse_alias(self._parse_conjunction())
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 3235, in _parse_conjunction
    return self._parse_tokens(self._parse_equality, self.CONJUNCTION)
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 4905, in _parse_tokens
    this = parse_method()
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 3238, in _parse_equality
    return self._parse_tokens(self._parse_comparison, self.EQUALITY)
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 4905, in _parse_tokens
    this = parse_method()
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 3241, in _parse_comparison
    return self._parse_tokens(self._parse_range, self.COMPARISON)
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 4905, in _parse_tokens
    this = parse_method()
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 3248, in _parse_range
    expression = self.RANGE_PARSERS[self._prev.token_type](self, this)
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 626, in <lambda>
    TokenType.FOR: lambda self, this: self._parse_comprehension(this),
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 5303, in _parse_comprehension
    return self.expression(
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 1117, in expression
    return self.validate_expression(instance)
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 1137, in validate_expression
    self.raise_error(error_message)
  File "/.pyenv/versions/3.10.6/lib/python3.10/site-packages/sqlglot/parser.py", line 1097, in raise_error
    raise error
sqlglot.errors.ParseError: Required keyword: 'this' missing for <class 'sqlglot.expressions.Comprehension'>. Line 1, Col: 47.
  FOR record IN (SELECT word FROM shakespeare) DO SELECT record.word; END FOR;
</details>

The syntax is described in the BigQuery docs [here](https://cloud.google.com/bigquery/docs/reference/standard-sql/procedural-language#for-in).

### PR DESCRIPTION
**Pull request by GeorgeSittas**
Fixes #2506

Reference: https://cloud.google.com/bigquery/docs/reference/standard-sql/procedural-language#for-in

### CODE REVIEW

File: sqlglot/dialects/bigquery.py [hunk 1]
```diff
@@ -358,6 +358,17 @@ class Parser(parser.Parser):
...
+        STATEMENT_PARSERS = {
+            **parser.Parser.STATEMENT_PARSERS,
+            TokenType.END: lambda self: self._parse_as_command(self._prev),
+            TokenType.FOR: lambda self: self._parse_for(),
+        }
+
+        def _parse_for(self) -> exp.ForIn:
+            this = self._parse_range()
+            self._match_text_seq("DO")
+            return self.expression(exp.ForIn, this=this, expression=self._parse_statement())
```
Comment (tobymao) [inline sqlglot/dialects/bigquery.py:370]: can you do multiple statements separated by semi colon?

File: sqlglot/dialects/bigquery.py [hunk 2]
```diff
@@ -358,6 +358,17 @@ class Parser(parser.Parser):
 
         NULL_TOKENS = {TokenType.NULL, TokenType.UNKNOWN}
 
+        STATEMENT_PARSERS = {
+            **parser.Parser.STATEMENT_PARSERS,
+            TokenType.END: lambda self: self._parse_as_command(self._prev),
+            TokenType.FOR: lambda self: self._parse_for(),
```
Comment (tobymao) [inline sqlglot/dialects/bigquery.py:364]: parse_for_in

File: sqlglot/dialects/bigquery.py [hunk 3]
```diff
@@ -358,6 +358,17 @@ class Parser(parser.Parser):
 
         NULL_TOKENS = {TokenType.NULL, TokenType.UNKNOWN}
 
+        STATEMENT_PARSERS = {
+            **parser.Parser.STATEMENT_PARSERS,
+            TokenType.END: lambda self: self._parse_as_command(self._prev),
```
Comment (tobymao) [inline sqlglot/dialects/bigquery.py:363]: why do you have this? shouldn't this be a part of the for in parser?

File: sqlglot/dialects/bigquery.py (hunk 1 above)
Comment (GeorgeSittas) [reply sqlglot/dialects/bigquery.py:370]: we split statements into chunks on `;` at the top level `_parse` method so it's not trivial

Comment (GeorgeSittas) [reply sqlglot/dialects/bigquery.py:370]: this should also answer:

> why do you have this? shouldn't this be a part of the for in parser?

File: sqlglot/dialects/bigquery.py (hunk 2 above)
Comment (GeorgeSittas) [reply sqlglot/dialects/bigquery.py]: check

**Pull request merged**